src/main/java/org/mule/extensions/vm/internal/listener/VMListener.java

```diff
@@ -0,0 +1,226 @@
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+
+/**
+ * A source which creates and listens on a VM queues.
+ *
+ * VM queues are created by placing listeners on them, which is why this listener contains parameters on the queue's
+ * behaviour, such as it being persistent or not, the max capacity, etc.
+ *
+ * The VM connector can only be used to publish and consume messages from queues for which a listener has been defined.
+ *
+ * @since 1.0
+ */
+@Alias("listener")
+@EmitsResponse
+public class VMListener extends Source<Serializable, VMMessageAttributes> {
+
+  private static final Logger LOGGER = getLogger(VMListener.class);
+  private static final String REPLY_TO_QUEUE_NAME = "replyTo";
+
+  @Inject
+  private VMConnectorQueueManager connectorQueueManager;
+
+  @ParameterGroup(name = "queue")
+  private QueueListenerDescriptor queueDescriptor;
+
+  /**
+   * The amount of concurrent consumers to be placed on the queue. As the number of consumers increases,
+   * so does the speed on which this source pushes messages into the owning flow. 
+   */
+  @Parameter
+  @Optional(defaultValue = "4")
+  @Alias("consumers")
+  private int consumersCount;
+
+  @Connection
+  private VMConnection connection;
+
+  @Inject
+  private SchedulerService schedulerService;
+
+  private List<Consumer> consumers;
+
+  private Scheduler scheduler;
+
+  private FlowInfo flowInfo;
+
+  @Override
+  public void onStart(SourceCallback<Serializable, VMMessageAttributes> sourceCallback) throws MuleException {
+    connectorQueueManager.createQueue(queueDescriptor, flowInfo.getName());
+    startConsumers(sourceCallback);
+  }
+
+  @Override
+  public void onStop() {
+    if (consumers != null) {
+      consumers.forEach(Consumer::stop);
+    }
+
+    if (scheduler != null) {
+      scheduler.stop();
+    }
+  }
+
+  @OnSuccess
+  public void onSuccess(@ParameterGroup(name = "Response", showInDsl = true) VMResponseBuilder messageBuilder,
+                        SourceCallbackContext ctx) {
+
+    ctx.<String>getVariable(REPLY_TO_QUEUE_NAME).ifPresent(replyTo -> {
+      Queue queue;
+      try {
+        queue = connection.getQueue(replyTo);
+      } catch (Exception e) {
+        LOGGER.warn(format("Found exception trying to obtain replyTo queue '%s'", replyTo), e);
+        return;
+      }
+
+      if (queue != null) {
+        try {
+          queue.offer(messageBuilder.getBody(), queueDescriptor.getQueueTimeoutInMillis());
+        } catch (Exception e) {
+          LOGGER.warn(format("Found exception trying to send response to replyTo queue '%s'", replyTo), e);
+        }
+      } else {
+        LOGGER.warn("Could not send response to replyTo queue '{}' because it does not exists", replyTo);
+      }
+    });
+  }
+
+  @OnTerminate
+  public void onTerminate() {
+    // no - op
```

Comment:
SDK forces you too when you have an @OnSuccess or @OnError. It will no longer be empty when the Transaction issue is fixed